storage/block/block.go

```diff
@@ -239,11 +241,19 @@ func (b *dbBlock) IsCachedBlock() bool {
 	return !retrieved || wasRetrieved
 }
 
-func (b *dbBlock) Merge(other DatabaseBlock) {
+func (b *dbBlock) Merge(other DatabaseBlock) error {
 	b.Lock()
+	if b.wasRetrievedFromDisk || other.WasRetrievedFromDisk() {
+		// We use Merge to lazily merge blocks that eventually need to be flushed to disk
+		// If we try to perform a merge on blocks that were retrieved from disk then we've
+		// violated an invariant and probably have a bug that is causing data loss.
+		b.Unlock()
+		return errTriedToMergeBlockFromDisk
+	}
 	b.resetMergeTargetWithLock()
 	b.mergeTarget = other
 	b.Unlock()
+	return nil
```

Comment:
Hm maybe we should double check here that `b.mergeTarget == nil` before setting the merge target too?  Otherwise that would also cause data loss.

Reply:
Actually in that case may we want to do `b.mergeTarget.Merge(other)`?